### PR TITLE
Support externals

### PR DESCRIPTION
**Pull request by pskd73**

### CODE REVIEW

File: pyluca/action.py [hunk 1]
```diff
@@ -84,12 +85,24 @@ def _apply_action(
...
+    elif action_type.startswith('external.'):
+        name = action_type.replace('external.', '')
+        args = {k: _get_param(v, event, accountant, context) for k, v in action.get('context', {}).items()}
+        externals[name](**args)
+    else:
+        raise NotImplementedError(f'"{action_type}" is not a valid action type!')
 
 
-def apply(event: Event, accountant: Accountant, context: dict = None):
+def apply(event: Event, accountant: Accountant, context: dict = None, externals: dict = None):
```
Comment (sattyamjjain) [inline pyluca/action.py:102]: instead of None, can we set {}

File: pyluca/action.py [hunk 2]
```diff
@@ -84,12 +85,24 @@ def _apply_action(
...
     context = context if context else {}
     event_config = accountant.config['actions_config']['on_event'][event.__class__.__name__]
     common_actions = accountant.config['actions_config'].get('common_actions', {})
+    externals = externals if externals else {}
```
Comment (sattyamjjain) [inline pyluca/action.py:106]: then we dont need this

**Review by sattyamjjain — CHANGES_REQUESTED** [review]

File: pyluca/action.py (hunk 1 above)
Comment (sattyamjjain) [reply pyluca/action.py:102]: can we use any other name variable,
externals is not giving the actual context

**Review by sattyamjjain — CHANGES_REQUESTED** [review]

File: pyluca/action.py (hunk 1 above)
Comment (pskd73) [reply pyluca/action.py]: PyCharm suggests not to use mutable as a default value.

File: pyluca/action.py [hunk 3]
```diff
@@ -84,12 +85,24 @@ def _apply_action(
...
-            _apply_action(sub_action, event, accountant, {**context, **action.get('context', {})}, common_actions)
+            _apply_action(
+                sub_action, event, accountant,
+                {**context, **action.get('context', {})},
+                common_actions,
+                external_actions
+            )
+    elif action_type.startswith('external_action.'):
+        name = action_type.replace('external_action.', '')
+        args = {k: _get_param(v, event, accountant, context) for k, v in action.get('context', {}).items()}
```
Comment (sameer-639) [inline pyluca/action.py:96]: instead of args make it kwargs

**Review by sattyamjjain — DISMISSED** [review]

**Review by ranjanikrrao — APPROVED** [review]

**Review by sameer-639 — APPROVED** [review]

**Review by sameer-639 — APPROVED** [review]

File: pyluca/action.py [hunk 4]
```diff
@@ -84,12 +85,23 @@ def _apply_action(
...
+                sub_action, event, accountant,
+                {**context, **action.get('context', {})},
+                common_actions,
+                external_actions
+            )
+    elif action_type.startswith('external_action.'):
+        kwargs = {k: _get_param(v, event, accountant, context) for k, v in action.get('context', {}).items()}
+        external_actions[action_type.replace('external_action.', '')](**kwargs)
+    else:
+        raise NotImplementedError(f'"{action_type}" is not a valid action type!')
```
Comment (ranjanikrrao) [inline pyluca/action.py:98]: Why double quotes again?

Comment (pskd73) [reply pyluca/action.py:98]: It is double quotes inside string

Comment (ranjanikrrao) [reply pyluca/action.py:98]: Okay

**Review by ranjanikrrao — APPROVED** [review]

**Review by sattyamjjain — APPROVED** [review]

**Review by adithyadinesh96 — APPROVED** [review]

**Pull request merged**